### ISSUE TITLE
[Feature request] Workspace switcher: implement ext-workspace protocol

### ISSUE THREAD

**Issue by stefonarch** [lxqt/lxqt-panel#2137]
As `cosmic-workspace-unstable-v1` is [merged](https://github.com/labwc/labwc/pull/2030) in labwc and will  be in 0.8.1 a labwc backend could be created, enabling the workspace switcher.

EDIT: fixed link

**Comment by isf63** [comment]
I recently investigated Wayland and found: https://wayland.app/protocols/

I think we shouldn't be too broad with supporting protocols.

**Comment by stefonarch** [comment]
Reality is: waybar and sfwbar (screenshot below) do it already, and as we support already the correspondent KDE protocol https://github.com/lxqt/lxqt-panel/tree/master/panel/backends/wayland/protocols I see no reason for excluding it as it will add functionality for labwc.

From the (fixed) link:
>but version 1 of the [cosmic-workspaces protocol](https://github.com/pop-os/cosmic-protocols/blob/main/unstable/cosmic-workspace-unstable-v1.xml) is basically the same as what is proposed in the upstream protocol.

![image](https://github.com/user-attachments/assets/18340592-8b49-48e9-815c-b82e8f3ea5c5)

**Comment by marcusbritanicus** [comment]
If waybar and swfbar do it, then it should be quite easy. We can use their code as a reference. If the license is compatible, then we might be able to borrow it (with attribution, of course) directly.

**Comment by stefonarch** [comment]
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/commit/605436b591729f70fb319fafbe3414657696841d

Labwc is implementing it already, don't know others, Hyprland dev isn't enthusiastic about it.

**Comment by stefonarch** [comment]
Niri is now the second compositor besides labwc implementing it:
https://github.com/YaLTeR/niri/pull/1800
@marcusbritanicus Any time to look into it?

**Comment by marcusbritanicus** [comment]
@stefonarch Sure. Let's do it sometime in the coming weeks.

**Comment by Jogai** [comment]
Would this also take the taskbar `limitByDesktop` setting into account?

**Comment by stefonarch** [comment]
> Would this also take the taskbar `limitByDesktop` setting into account?

No, this would need a IPC implementation similar of the one for Wayfire which is WIP; including move to desktop in right click menu.

**Comment by stefonarch** [comment]
It's [merged](https://github.com/YaLTeR/niri/commit/ce9ba00d54104f250f0074a54f14d1a5e4ff427f) in master Niri, and also xfce has fixed apparently multimonitor pager issues. @marcusbritanicus any free time?

**Comment by marcusbritanicus** [comment]
@stefonarch Thanks for the update. Let's see how we can implement it. I will have to start out by installing niri first.

**Comment by stefonarch** [comment]
> [@stefonarch](https://github.com/stefonarch) Thanks for the update. Let's see how we can implement it. I will have to start out by installing niri first.

Labwc is fine too, but doesn't have per-output-workspaces (where xfce4-panel stumbled hard, should be fixed now in git).

**Comment by tsujan** [comment]
> Let's see how we can implement it.

Looking forward to your work. As usual, I'll appear when you write a code, because only then I'll know how things are done ;)

**Comment by marcusbritanicus** [comment]
@stefonarch #2322 I got some time yesterday and today, so made a little progress. The classes themselves are quite simple and easy to use. 

If you, or @tsujan or @gfgit have some time, please feel free to work on this. Feel free to open a PR against my branch. Or you can always fork my repo and continue my work. I do not mind either.
